Fixes and simplifications from static analysis

awemany opened the pull request.

kallewoof committed:
[zmq] Call va_end() on va_start()ed args.

awemany commented:
Don't know why travis stopped with a failure. @sickpig, can you have a look? Thank you :-)

awemany commented:
@carmay: Ooops. I *just* updated them (also removed a couple dead assignments) and squashed everything into a single commit. Sorry for that. Can you have another look? :)

sickpig commented on src/leveldb/db/log_reader.cc:
```diff
@@ -39,7 +39,6 @@ bool Reader::SkipToInitialBlock() {
 
   // Don't search a block if we'd be in the trailer
   if (offset_in_block > kBlockSize - 6) {
-    offset_in_block = 0;
```
leveldb is maintained as subtree, we usually don't change leveldb locally. Better way to do it is to submit a patch upstream or check if upstream has a fix already.

awemany replied:
Fair point, I'll have a look and take the level DB changes out of this commit.

sickpig commented on src/secp256k1/src/ecmult_gen_impl.h:
```diff
@@ -150,7 +150,10 @@ static void secp256k1_ecmult_gen(const secp256k1_ecmult_gen_context *ctx, secp25
     }
+    /* see also: https://github.com/bitcoin-core/secp256k1/pull/485 */
+#ifndef __clang_analyzer__
     bits = 0;
```
secp256k1 is maintained via subtree

awemany replied:
Makes sense, however notice that the referenced PR basically says that libsecp256k1 is not yet updated either. So this is just meant to quieten the static analyzer.

awemany replied:
Ok, I addressed this now with a separate copy of the diff in a `patch/` subdirectory.

awemany commented:
Travis looks like a spurious failure.

awemany commented:
Failure was spurious, I took out the `leveldb` changes.

awemany committed:
Prevent uninitialized members in CMutex(Read)Lock

(from clang static analysis)

awemany committed:
tx_invalid test case: Fix uninitialized variable

(from clang static analysis)

awemany committed:
wallet/crypto_tests: Fix memory_cleanse() invocation

(from clang static analysis)

awemany committed:
CRequestManager::Rejected() remove redundancy

Remove redundant check for REJECT_NONSTANDARD.

(from clang static analysis)

awemany committed:
serialize: Prevent uninitialized data when reading streams

Currently, reads past the end of a stream will produce undefined
results in serialize.h. This fixes (at least some of) the unitialized
reads.

awemany committed:
requestManager: Fix potential division by zero

This makes the calculation of nOverallAverageResponseTime more robust.
It looks like this can only triggered in the unlikely condition that
nMaxOutConnections is set to zero.

(from clang static analysis)

awemany committed:
getblockchaininfo: Fix case of pruning enabled but not pruned yet

This skips adding the 'pruneheight' entry when the blockchain is
marked as pruned but no blocks have been pruned yet. This might be a
state that is never reached.

(from clang static analysis)

awemany commented:
Ok, I went through the more thorough analysis and added a couple commits (one ported from Core) to fix further issues found by static analysis (as well as doing a rebase).

@sickpig: Regarding the code to make the SA quiet in the libsecp256k1 code, I can take that out if you want so that we keep a clean upstream repo -? Should I make it all separate commits or a single big one?

ptschip commented on src/wallet/wallet.cpp:
```diff
@@ -3323,7 +3323,7 @@ bool CWallet::InitLoadWallet()
 
-    CBlockIndex *pindexRescan = chainActive.Tip();
+    CBlockIndex *pindexRescan = nullptr;
     if (GetBoolArg("-rescan", false))
```
this looks to me like you're changing the functionality of rescan and will cause more rescans on startup than is necessary.

awemany replied:
Hmm? As far as I can see, `pindexRescan` is initialized in all codepaths below and thus this one seems superfluous - and was detected by clang SA as such.

Or are you intending a side-effect of calling `chainActive.Tip()` here? As far as I can see, that method also simply returns a value.

ptschip replied:
I was looking at line 3338.  But you're right, pindexRescan by the time it gets to 3338 will have been assigned another value so it's fine.

awemany committed:
Fix a couple issues found through clang static analysis

In particular, this fixes:
 - a couple dead initializations and assignments in the main code base
 - simplify GetChance() in addrman.cpp
 - simplify pindexRescan initialization in wallet.cpp
 - use __clang_analyzer__ #ifndefs to hide a dead increment in pubkey.cpp (plus comments)

awemany committed:
Fix a 'never read' static analysis warning in libsecp256k1

See also:

https://github.com/bitcoin-core/secp256k1/pull/485

This also creates a directory for post-import of 3rdparty library
patches and populates it with a diff of this change.

awemany commented:
Ok, I rebased everything and reworked the one touching `libsecp256k1` to include a separate patch in a newly created `patches/` directory.

sickpig commented on src/sync.h:
```diff
@@ -328,10 +328,10 @@ class SCOPED_LOCKABLE CMutexReadLock
+    uint64_t lockedTime = 0;
+    const char *name = "unknown-name";
+    const char *file = "unknown-file";
+    int line = 0;
```
could `line` become negative? if this is not the case I wonder we could change is type to uint, no?

awemany replied:
Good point. I don't think any lock is created anywhere with a negative line number, but I'll check just to be sure. 

However, whether it could become negative also depends on the behavior of expanding `__LINE__` and thus potentially implementation details of the compiler's preprocessor.

Are we sure compilers will never produce a negative value for `__LINE__`, for example in the hypothetical scenario of compiling `stdin`?

I just tested both `gcc`'s `cpp` as well as `clang`'s `cpp` and they seem to start at line no 1 from `stdin`, as expected.

There's also no arithmetic done with the line numbers so I guess we should be safe. I'll update the code.

sickpig replied:
great

awemany replied:
@sickpig: There's a lot of functions that take line numbers as an `int` in `sync.h`. Should I change all of them to `unsigned`?

gandrewstone replied:
sure, makes sense to me

sickpig commented on src/requestManager.cpp:
```diff
@@ -360,9 +360,6 @@ void CRequestManager::Rejected(const CInv &obj, CNode *from, unsigned char reaso
     }
-    else if (reason == REJECT_NONSTANDARD)
-    {
-    }
```
wonder if this is intended as a sort of placeholder for future cases needed to handle. @ptschip, @gandrewstone?

sickpig replied:
scratch that didn't notice the we check for that state a few lines above

gandrewstone replied:
probably an auto merge error

sickpig commented on src/net.cpp:
```diff
@@ -1510,7 +1510,6 @@ static std::string GetDNSHost(const CDNSSeedData &data, uint64_t requiredService
     // use default host for non-filter-capable seeds or if we use the default service bits (NODE_NETWORK)
     if (!data.supportsServiceBitsFiltering || requiredServiceBits == NODE_NETWORK)
     {
-        requiredServiceBits = NODE_NETWORK;
```
correct. cause `requiredServiceBits` to NODE_NETWORK just before calling `GetDNSHost()` at line 1571

awemany commented:
Took this out of WIP, ready for review. Note that there was a weird and interesting travis failure that unfortunately got lost as a redo of a travis job seems to be done under the old job label.

gandrewstone commented on src/netbase.cpp:
```diff
@@ -756,7 +756,7 @@ std::string NetworkErrorString(int err)
+    const char *s = nullptr;
     buf[0] = 0;
 /* Too bad there are two incompatible implementations of the
  * thread-safe strerror. */
```
If the POSIX path is taken, s is still set as nullptr so the strprintf() show the error.
I think you need "s = buf" in the POSIX path

awemany replied:
Indeed! I changed it so that it is properly set in the non-POSIX path but explicitly only in that code path.

gandrewstone commented on src/requestManager.cpp:
```diff
@@ -1097,7 +1094,14 @@ bool CRequestManager::MarkBlockAsReceived(const uint256 &hash, CNode *pnode)
+                        LOG(IBD, "Calculation of average response time failed due to division by zero.\n");
+                    }
                 }
                 nOverallAverageResponseTime += nResponseTime / nOverallRange;
```
here's another potential / 0.  maybe wrap the entire code in one if check?

awemany replied:
Good catch, have done so now. Accidentally clicked 'cancel my review' button, I don't know what that did. All/most my comments still seem to be here?

awemany commented:
It just came to my mind that I should also note that

 `serialize: Prevent uninitialized data when reading streams` 

is not from clang static analysis (I think I have seen this in an earlier static analysis, though) and that maybe one should think about scenarios where this might degrade performance. I cannot see where it could, though, and I think clearing the variable before reading is very likely better than processing uninitialized memory.
